### PR TITLE
 feat(assertion) add resolves/rejects assertions

### PR DESCRIPTION
**Pull request by mgred**
Copy of #44 because I accidentally closed with `git push --force` :man_facepalming:  

#### Purpose

This is a first attempt for #27 

#### Solution

Adds a `resolves` assertion to **referee**, that uses `samsam`'s `identical` method to check if a promise resolves to an expected value. The assertion returns a Promise which can be passed to the test runner to know that this is an async method.

* If both values are identical the returned Promise resolves.
* If both values are different the returned Promise rejects and fails.
* If the given Promise/async method does not resolve, the returned Promise will reject.
* If the given input is not a Promise/async method the, the returned Promise will reject

#### How to verify - mandatory
1. Check out this branch
2. `npm install`
3. `npm test`

#### Checklist for author

- [x] `npm run lint` passes
- [x] References to standard library functions are [cached](https://github.com/sinonjs/sinon/pull/1523).
- [ ] Write sufficient tests
- [x] Add documentation

### CODE REVIEW

**Comment by mroderick** [comment]
Thank you @mgred for looking into this one.

I tried out the branch with one concern that I have: because of the async nature, it's very easy to making passing tests that shouldn't actually pass.

```js
"use strict";

var assert = require("./lib/referee").assert;

function myAsyncFunction() {
    return Promise.resolve(42);
}

describe("test resolves", function() {
    it("should something", function() {
        assert.resolves(myAsyncFunction(), 43);
    });
});
```

If you try running this test, it will pass. However if you read it carefully, I think you will agree that it should not pass.

I am wondering if having this kind of feature, will make it more likely that people make tests that give them a false sense of security, doing them a disservice by allowing bugs to go undetected.

Another thing I've been thinking about is, can we somehow detect incorrect usage of this functionality? Would it be helpful to make an ESLint plugin for `referee`, that would detect when the author is not doing anything with the promise returned (i.e. not returning the promise or not prefixing it with the `await` keyword).

**Comment by mgred** [comment]
> If you try running this test, it will pass. However if you read it carefully, I think you will agree that it should not pass.
> 
> I am wondering if having this kind of feature, will make it more likely that people make tests that give them a false sense of security, doing them a disservice by allowing bugs to go undetected.

Yeah, see your point. I think we have to make this clear in the documentation, that this assertion needs to be returned to be properly handled by the test runner.

> Another thing I've been thinking about is, can we somehow detect incorrect usage of this functionality? Would it be helpful to make an ESLint plugin for `referee`, that would detect when the author is not doing anything with the promise returned (i.e. not returning the promise or not prefixing it with the `await` keyword).

That's a nice idea and could help to prevent that mistake.
I can imagine it might be easy to determine in case of your example
```js
describe("test resolves", function() {
    it("should something", function() {
         // eslint complains since `assert.resolve` is used without to be returned
        assert.resolves(myAsyncFunction(), 43);
    });
});
```

But could that work with e.g.:
```js
const { resolve } = assert;
describe("test resolves", function() {
    it("should something", function() {
        resolves(myAsyncFunction(), 43);
    });
});
```

**Comment by mroderick** [comment]
I think it should be possible to trace back the origin of the `resolves` and `refutes` to `referee`, so that we only examine actual `referee` methods, and not just any function called `resolves` or `refutes`.

Maybe it would also be possible to detect unhandled promises and put that in `referee.verify()` (see #48)

**Comment by mgred** [comment]
> I think it should be possible to trace back the origin of the `resolves` and `refutes` to `referee`, so that we only examine actual `referee` methods, and not just any function called `resolves` or `refutes`.

:+1: 

> Maybe it would also be possible to detect unhandled promises and put that in `referee.verify()` (see #48)

Sounds great, would be my option of choice so far :smile:

**Comment by mgred** [comment]
I think from that point, we can start refactoring this to not have all the code duplication

**Comment by mroderick** [comment]
If I can offer some advice: before you start refactoring, maybe you want to clean up the history first. Doing that while the code still looks mostly the same is going to make it easier to recognise things, than after you've made considerable changes to the structure

File: lib/assertions/rejects.test.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+"use strict";
+
+var referee = require("../referee");
+
+describe("assert.rejects", function() {
+    it("should resolve and pass when identical", function() {
```
[the remark this answers is not in the document]
Comment (mgred) [reply lib/assertions/rejects.test.js]: Thanks for your review.
Just to make sure I got you correctly. Do you means something like this:
```js
describe("assert.rejects", function() {
    it("should return a promise", function() {
        var assertion = referee.assert.rejects(Promise.reject("test"), "test");
        referee.assert.isPromise(assertion);
    });

    context("when promise argument rejects to value argument", function() {
        it("should resolve the returned promise", function() {
            return referee.assert.rejects(Promise.reject("test"), "test");
        });
    });

    context(
        "when promise argument does not reject to value argument",
        function() {
            it("should reject the returned promise", function() {
                return referee.assert
                    .rejects(Promise.reject("test"), "test2")
                    .catch(function(e) {
                        referee.assert.isError(e);
                    });
            });
        }
    );

    context("when promise argument is not a promise", function() {
        it("should reject the returned promise", function() {
            return referee.assert.rejects({}, "test").catch(function(e) {
                referee.assert.isError(e);
            });
        });
    });

    context("when promise argument does not reject", function() {
        it("should reject the returned promise", function() {
            return referee.assert
                .rejects(Promise.resolve(), "test")
                .catch(function(e) {
                    referee.assert.isError(e);
                });
        });
    });
});
```

[the remark this answers is not in the document]
Comment (mroderick) [reply lib/assertions/rejects.test.js]: Yes :)

[the remark this answers is not in the document]
Comment (mgred) [reply lib/assertions/rejects.test.js]: Cool, I will then refactor all other tests that way :+1: 
Thanks a lot

File: docs/index.md [hunk 2]
```diff
@@ -1455,6 +1460,73 @@ assert.className(el, "feed items");     // Fails, "items" is not a match
 assert.className(el, ["item", "feed"]); // Passes
 ```
 
+### `resolves()`
+
+```js
+assert.resolves(Promise, value)
```
[the remark this answers is not in the document]
Comment (mgred) [reply docs/index.md]: Yeah, that's indeed a bit hairy :smile: I will rename it to `promise`, also for the rest
Thanks for your feedback!

**Pull request merged**